File: src/main/java/scrabble/phrases/providers/DefinitionProvider.java
package scrabble.phrases.providers;

import scrabble.phrases.dictionary.WordDictionary;
import scrabble.phrases.words.Adjective;
import scrabble.phrases.words.Noun;
import scrabble.phrases.words.NounGender;
import scrabble.phrases.words.Verb;

/**
 * Provides dictionary-style definition sentences.
 * Format: NOUN: [adj] [noun] care [verb].
 */
public class DefinitionProvider extends SentenceProvider {

    public DefinitionProvider(WordDictionary dictionary) {
        super(dictionary);
    }

    @Override
    protected void initFilters() {
        // No filters needed
    }

    @Override
    public String getSentence() {
        Noun defined = getDictionary().getRandomNoun();
        Noun noun = getDictionary().getRandomNoun();
        Adjective adj = getDictionary().getRandomAdjective();
        Verb verb = getDictionary().getRandomVerb();
        if (defined == null || noun == null || adj == null || verb == null) {
            throw new IllegalStateException("Dictionary has insufficient words for definition");
        }

        String adjForm = (noun.gender() == NounGender.F) ? adj.feminine() : adj.word();
        return defined.word().toUpperCase() + ": " + noun.articulated() + " " + adjForm + " care " + verb.word() + ".";
    }
}

File: src/main/kotlin/scrabble/phrases/providers/DefinitionProvider.kt
package scrabble.phrases.providers

import scrabble.phrases.repository.WordRepository
import scrabble.phrases.words.NounGender

class DefinitionProvider(
    private val repo: WordRepository,
    private val minRarity: Int = 1,
    private val maxRarity: Int
) : ISentenceProvider {

    override fun getSentence(): String {
        try {
            val defined = repo.getRandomNoun(minRarity = minRarity, maxRarity = maxRarity)
            val noun = repo.getRandomNoun(minRarity = minRarity, maxRarity = maxRarity, exclude = setOf(defined.word))
            require(noun.gender == NounGender.M || noun.gender == NounGender.F) { "noun '${noun.word}' has invalid gender $noun.gender" }
            val adj = repo.getRandomAdjective(minRarity = minRarity, maxRarity = maxRarity)
            val verb = repo.getRandomVerb(minRarity = minRarity, maxRarity = maxRarity)
            val object_ = repo.getRandomNoun(minRarity = minRarity, maxRarity = maxRarity, exclude = setOf(defined.word, noun.word))

            return "${defined.word.uppercase()}: ${noun.articulated} ${adj.forGender(noun.gender)} care ${verb.word} ${object_.articulated}."
        } catch (e: IllegalStateException) {
            throw IllegalStateException(
                "DefinitionProvider needs at least one noun, adjective and verb in rarity range ${rarityDesc(minRarity, maxRarity)} — database may be empty or misconfigured", e
            )
        }
    }

    private fun rarityDesc(min: Int, max: Int): String =
        if (min <= 1) "<= $max" else "between $min and $max"
}
